tools/flowc/backends/nim/runtime.nim

```diff
@@ -45,48 +47,139 @@ proc rt_to_double*(x: float): float = x
 proc rt_to_double*(x: bool): float = return if x: 1.0 else: 0.0
 proc rt_to_double*(x: string): float = parseFloat(x)
 
-#[ Runtime type ids ]#
-
 type
+  # Basic runtime type kinds
   RtType* = enum
-    rtVoid,
-    rtBool,
-    rtInt,
-    rtDouble,
-    rtString,
-    rtNative,
-    rtArray,
-    rtFunc,
-    rtStruct
+    # Atiomic types
+    rtVoid, rtBool, rtInt, rtDouble, rtString, rtNative,
+    # Composite types
+    rtArray, rtFunc, rtStruct
 
-#[ Representation of a dynamic type ]#
+  # Compile time type kinds
+  CtType* = enum
+    # Atiomic types
+    ctVoid, ctBool, ctInt, ctDouble, ctString, ctNative, ctFlow, ctUnion,
+    # Composite types
+    ctArray, ctFunc, ctStruct
 
-  FlowField* = object
-    name*: string
-    val*: Flow
+  # A complex type descriptor.
+  AlType* = tuple[op: CtType, args: seq[int32], name: string]
+
+#[ Representation of a dynamic type ]#
 
   Flow* = ref object
     case tp*: RtType
+    # Atiomic types
     of rtVoid:   discard
     of rtBool:   bool_v:   bool
     of rtInt:    int_v:    int32
     of rtDouble: double_v: float
     of rtString: string_v: string
     of rtNative: native_v: Native
+    # Composite types
     of rtArray:  array_v:  seq[Flow]
     of rtFunc:   func_v:   proc(x: seq[Flow]): Flow
     of rtStruct:
       str_id: int32
       str_name: string
-      str_fields: seq[FlowField]
+      str_args: seq[Flow]
 
   Struct* = ref object of RootObj
     id: int32
+  Array* = ref object
+    id: int32
 
   Native* = ref object of RootObj
     what: string
     val: RootObj
 
+# Type index oprations
+var id2type*: seq[AlType]
+var type2id*: Table[AlType, int32]
+var id2fields*: seq[seq[string]]
+var struct2id*: Table[string, int32]
+
+proc rt_type_id_to_string*(id: int32): string =
+  let tp = id2type[id]
+  case tp.op:
+  of ctArray:  return "[" & rt_type_id_to_string(tp.args[0]) & "]"
+  of ctFunc:   return "(" & map(tp.args[1..tp.args.len - 1], proc (arg: int32): string = rt_type_id_to_string(arg)).join(", ") & ") -> " & rt_type_id_to_string(tp.args[0])
+  of ctStruct: return tp.name & "(" & map(tp.args[1..tp.args.len - 1], proc (arg: int32): string = rt_type_id_to_string(arg)).join(", ") & ")"
+  else: tp.name
+
+proc rt_type_id_to_struct_id*(id: int32): int32 =
+  return id2type[id].args[0]
+
+proc rt_find_type_id*(tp: AlType): int32 =
+  return if not type2id.hasKey(tp): -1 else: type2id[tp]
```

Comment:
If a type2id hash table does **not** have a type, it returns -1, otherwise it returns its index.  What's wrong?
Maybe you mean that 
`return if type2id.hasKey(tp): type2id[tp] else: -1`
would be more readable?
ok, accepted.